hera_cal/io.py

```diff
@@ -7,26 +7,537 @@
+                      for antpair in antpairs}
+
+        locs = locals()
+        return {meta: eval(meta, {}, locs) for meta in self.HERAData_metas}
+
+    def _determine_blt_slicing(self):
+        '''Determine the mapping between antenna pairs and
+        slices of the blt axis of the data_array.'''
+        self._blt_slices = {}
+        for ant1, ant2 in self.get_antpairs():
+            indices = self.antpair2ind(ant1, ant2)
+            if len(indices) == 1:
+                self._blt_slices[(ant1, ant2)] = slice(indices[0], indices[0] + 1, self.Nblts)
+            elif not (len(set(np.ediff1d(indices))) == 1):
+                raise NotImplementedError('UVData objects with non-regular spacing of '
+                                          'baselines in its baseline-times are not supported.')
+            else:
+                self._blt_slices[(ant1, ant2)] = slice(indices[0], indices[-1] + 1,
+                                                       indices[1] - indices[0])
+
+    def _determine_pol_indexing(self):
+        '''Determine the mapping between polnums and indices
+        in the polarization axis of the data_array.'''
+        self._polnum_indices = {}
+        for i, polnum in enumerate(self.polarization_array):
+            self._polnum_indices[polnum] = i
+
+    def _get_slice(self, data_array, key):
+        '''Return a copy of the Nint by Nfreq waterfall or waterfalls for a given key. Abstracts
+        away both baseline ordering (by applying complex conjugation) and polarization capitalization.
+
+        Arguments:
+            data_array: numpy array of shape (Nblts, 1, Nfreq, Npol)
+            key: if of the form (0,1,'xx'), return anumpy array.
+                 if of the form (0,1), return a dict mapping pol strings to waterfalls.
+                 if of of the form 'xx', return a dict mapping ant-pair tuples to waterfalls.
+        '''
+        if isinstance(key, str):  # asking for a pol
+            return {antpair: self._get_slice(data_array, antpair + (key,)) for antpair in self.get_antpairs()}
+        elif len(key) == 2:  # asking for antpair
+            pols = np.array([polnum2str(polnum) for polnum in self.polarization_array])
+            return {pol: self._get_slice(data_array, key + (pol,)) for pol in pols}
+        elif len(key) == 3:  # asking for bl-pol
+            try:
+                return np.array(data_array[self._blt_slices[tuple(key[0:2])], 0, :,
+                                           self._polnum_indices[polstr2num(key[2])]])
+            except KeyError:
+                return np.conj(data_array[self._blt_slices[tuple(key[1::-1])], 0, :,
+                                          self._polnum_indices[polstr2num(conj_pol(key[2]))]])
+        else:
+            raise KeyError('Unrecognized key type for slicing data.')
+
+    def _set_slice(self, data_array, key, value):
+        '''Update data_array with Nint by Nfreq waterfall(s). Abstracts away both baseline
```

Comment:
same here in terms of elaborating on what `data_array` should be

Reply:
Done.